src/Compilers/CSharp/Portable/Binder/PatternExplainer.cs

```diff
@@ -0,0 +1,512 @@
+                    }
+
+                    return "(" + string.Join(", ", subpatterns) + ")" + (subpatterns.Count == 1 ? " { }" : null);
+                }
+
+                return null;
+
+                static string makeConjunct(string oldPattern, string newPattern) => (oldPattern, newPattern) switch
+                {
+                    ("_", var x) => x,
+                    (var x, "_") => x,
+                    (var x, var y) => x + " and " + y
+                };
+            }
+
+            // Handle the special case of numeric limits
+            string tryHandleNumericLimits()
+            {
+                if (evaluations.IsEmpty &&
+                    constraints.All(t => t switch
+                    {
+                        (BoundDagValueTest _, _) => true,
+                        (BoundDagRelationalTest _, _) => true,
+                        (BoundDagExplicitNullTest _, false) => true,
+                        (BoundDagNonNullTest _, true) => true,
+                        _ => false
+                    }) &&
+                    ValueSetFactory.ForType(input.Type) is { } fac)
+                {
+                    // All we have are numeric constraints. Process them to compute a value not covered.
+                    var remainingValues = fac.AllValues;
+                    foreach (var constraint in constraints)
+                    {
+                        var (test, sense) = constraint;
+                        switch (test)
+                        {
+                            case BoundDagValueTest v:
+                                addRelation(BinaryOperatorKind.Equal, v.Value);
+                                break;
+                            case BoundDagRelationalTest r:
+                                addRelation(r.Relation, r.Value);
+                                break;
+                        }
+                        void addRelation(BinaryOperatorKind relation, ConstantValue value)
+                        {
+                            var filtered = fac.Related(relation, value);
+                            if (!sense)
+                                filtered = filtered.Complement();
+                            remainingValues = remainingValues.Intersect(filtered);
+                        }
+                    }
+
+                    if (remainingValues.Complement().IsEmpty)
+                        return "_";
+
+                    return SampleValueString(remainingValues, input.Type, requireExactType: requireExactType);
+                }
+
+                return null;
+            }
+
+            // Handle the special case of a recursive pattern
+            string tryHandleRecursivePattern()
+            {
+                if (constraints.IsEmpty && evaluations.IsEmpty)
+                    return null;
+
+                string typeName = requireExactType ? input.Type.ToDisplayString() : null;
+
+                foreach (var con in constraints)
```

Comment:
>foreach (var con in constraints) [](start = 16, length = 32)

Consider using: `if (constraints.All(c => switch (c) { ... }))` as in previous local function.

Reply:
I cannot do so because of the early return.

---
In reply to: [443077432](https://github.com/dotnet/roslyn/pull/44702#discussion_r443077432) [](ancestors = 443077432)